### ISSUE TITLE
Performance: azd up does re-runs similar commands over and over

### ISSUE THREAD

**Issue by davidfowl** [Azure/azure-dev#3303]
There's a lot of redundancy when azd up runs that we should fix it. It seems like every azd operation: 
1. Runs a check to see if IsAspireHost is true.
2. Tries to execute the application host to grab the manifest.

Deploy seems to:
1. Publish the project to a temp folder
2. Publish to a container registry

example.

azd package:

```
2024/02/06 21:54:06 command_runner.go:307: Run exec: 'dotnet msbuild C:\<redacted>\eShop.AppHost\eShop.AppHost.csproj --getProperty:IsAspireHost' , exit code: 0

c: 'dotnet run --project eShop.AppHost.csproj --publisher manifest --output-path C:\<redacted>\AppData\Local\Temp\azd-provision2997516452\apphost-manifest.json' , exit code: 0
```

azd provision

```
2024/02/06 21:54:21 command_runner.go:307: Run exec: 'dotnet msbuild C:\<redacted>\eShop.AppHost\eShop.AppHost.csproj --getProperty:IsAspireHost' , exit code: 0
Additional env:
   DOTNET_CLI_WORKLOAD_UPDATE_NOTIFY_DISABLE=<redacted>

c: 'dotnet run --project eShop.AppHost.csproj --publisher manifest --output-path C:\<redacted>\AppData\Local\Temp\azd-provision3448215641\apphost-manifest.json' , exit code: 0
```

azd deploy

```
2024/02/06 21:54:47 command_runner.go:307: Run exec: 'dotnet msbuild C:\<redacted>\eShop.AppHost\eShop.AppHost.csproj --getProperty:IsAspireHost' , exit code: 0
Additional env:
   DOTNET_CLI_WORKLOAD_UPDATE_NOTIFY_DISABLE=<redacted>
```

Each project is published to a temp folder (which isn't used), then published to the container registry.

```
2024/02/06 21:54:47 command_runner.go:307: Run exec: 'dotnet msbuild C:\<redacted>\eShop.AppHost\eShop.AppHost.csproj --getProperty:IsAspireHost' , exit code: 0
Additional env:
   DOTNET_CLI_WORKLOAD_UPDATE_NOTIFY_DISABLE=<redacted>
```

```
 |  =====| Deploying service identity-api (Pushing container image)2024/02/06 21:56:55 command_runner.go:307: Run exec: 'dotnet publish C:\<redacted>\Identity.API\Identity.API.csproj -r linux-x64 -c Debug -p:PublishProfile=DefaultContainer -p:ContainerImageName=azd-deploy-identity-api-1707285401 -p:ContainerRegistry=<redacted>' , exit code: 0
Additional env:
```

**Comment by weikanglim** [comment]
> Publish the project to a temp folder

This was unintended, fixed with #3301 

> Runs a check to see if IsAspireHost is true.

This is a side-effect of an earlier decision to have "AppHost" projects referenced as-if if they were `dotnet` projects, in `azure.yaml`:

```yaml
services:
  AppHost:
    project: ./AppHost/Apphost.csproj
    language: dotnet
```

`azd` doesn't currently store metadata about the `AppHost` being an actual "app host" project, so it ends up re-evaluating.

Perf cost (evaluated on my machine): ~1 second

> Tries to execute the application host to grab the manifest.

The manifest is hugely important to `azd` right now.
- In package/deploy, `azd` needs the manifest to understand what AppHost referenced projects, i.e. services exist within the solution, so that `azd deploy <service>` works.
- In provision, `azd` needs the manifest to understand inputs, connection strings, etc.

Perf cost (evaluated on my machine): ~30 seconds on Windows

I'm running some experiments here, and I think #3231 may be the right track. I am seeing subsequent runs being somewhat cached reduce cost to around ~6s. On *nix, it's ~3s.

**Comment by davidfowl** [comment]
It should be 0 as we don’t need the manifest this many times. We can also do work in aspire to make this faster
